pika/channel.py

```diff
@@ -539,41 +540,45 @@ def close(self, reply_code=0, reply_text="Normal shutdown"):
                   self._on_closeok, [spec.Channel.CloseOk])
 
-    def confirm_delivery(self, callback=None, nowait=False):
+    def confirm_delivery(self, ack_nack_callback, callback=None):
         """Turn on Confirm mode in the channel. Pass in a callback to be
         notified by the Broker when a message has been confirmed as received or
         rejected (Basic.Ack, Basic.Nack) from the broker to the publisher.
 
         For more information see:
             http://www.rabbitmq.com/extensions.html#confirms
 
-        :param callable callback: The callback for delivery confirmations that
-            has the following signature: callback(pika.frame.Method), where
-            method_frame contains either method `spec.Basic.Ack` or
-            `spec.Basic.Nack`.
-        :param bool nowait: Do not send a reply frame (Confirm.SelectOk)
+        :param callable ack_nack_callback: Required callback for delivery
+            confirmations that has the following signature:
+            callback(pika.frame.Method), where method_frame contains
+            either method `spec.Basic.Ack` or `spec.Basic.Nack`.
+        :param callable callback: Callback to call when receiving
+            Confirm.SelectOk
+        :raises ValueError:
 
         """
-        self._validate_channel_and_callback(callback)
+        if not callable(ack_nack_callback):
+            # NB: confirm_deliver requires a callback; it's meaningless
+            # without a user callback to receieve Basic.Ack/Basic.Nack notifications
+            raise ValueError('confirm_delivery requires a callback '
+                             'to receieve Basic.Ack/Basic.Nack notifications')
 
-        # TODO confirm_deliver should require a callback; it's meaningless
-        # without a user callback to receieve Basic.Ack/Basic.Nack notifications
+        self._validate_channel()
+        nowait = self._validate_rpc_completion_callback(callback)
 
         if not (self.connection.publisher_confirms and
                 self.connection.basic_nack):
             raise exceptions.MethodNotImplemented('Not Supported on Server')
 
-        # Add the ack and nack callbacks
-        if callback is not None:
-            self.callbacks.add(self.channel_number, spec.Basic.Ack, callback,
-                               False)
-            self.callbacks.add(self.channel_number, spec.Basic.Nack, callback,
-                               False)
+        # Add the ack and nack callback
+        self.callbacks.add(self.channel_number, spec.Basic.Ack, ack_nack_callback,
+                           False)
+        self.callbacks.add(self.channel_number, spec.Basic.Nack, ack_nack_callback,
+                           False)
 
-        # Send the RPC command
         self._rpc(spec.Confirm.Select(nowait),
-                  self._on_selectok if not nowait else None,
-                  [spec.Confirm.SelectOk] if nowait is False else [])
+                  callback if not nowait else None,
```

Comment:
Since callback being None or not None now determines the value of `nowait`, technically `callback if not nowait else None` can be reduced to just `callback`